README.md
# Puppet Operations Appliance

#### Table of Contents

- [Puppet Operations Appliance](#puppet-operations-appliance)
      - [Table of Contents](#table-of-contents)
  - [Description](#description)
  - [Setup](#setup)
    - [What Puppet_Operations_Appliance modifies in your PE Installation](#what-puppet_operations_appliance-modifies-in-your-pe-installation)
    - [Setup Requirements](#setup-requirements)
      - [Module Dependencies](#module-dependencies)
      - [Minimum Hardware requirements](#minimum-hardware-requirements)
      - [OS Restrictions](#os-restrictions)
    - [Beginning with Puppet_Operations_Appliance](#beginning-with-puppet_operations_appliance)
  - [Usage](#usage)
    - [Live Telemetry Display](#live-telemetry-display)
    - [Infrastructure node file and log access](#infrastructure-node-file-and-log-access)
      - [Optional Configuration](#optional-configuration)
    - [PE Client tools](#pe-client-tools)
      - [Creating Support User](#creating-support-user)
    - [Puppet Enterprise Database Access](#puppet-enterprise-database-access)
  - [Uninstallation](#uninstallation)
  - [Limitations](#limitations)
  - [Known Issues](#known-issues)
  - [Contributions](#contributions)
- [Supporting Content](#supporting-content)
    - [Articles](#articles)
    - [Videos](#videos)


## Description

The Puppet Operations Appliance is designed to be a central point to which a Puppet Enterprise environment may be monitored and maintained.
The appliance collates data and provides read-only access, useful in incident resolution and preventative maintenance.


## Setup

### What Puppet_Operations_Appliance modifies in your PE Installation 

Puppet_Operations_Appliance will Export NFS mounts of key directories from each Puppet Enterprise infrastructure node, while also setting up requirements for gathering of metrics and database access for the Puppet_Operations_Appliance.
Open-source software required for the proper functioning of the Puppet_Operations_Appliance will be deployed on the target agent node.


### Setup Requirements 

#### Module Dependencies

- derdanne/nfs (>= 2.1.5)
- puppetlabs/postgresql (>= 6.6.0)
- puppetlabs/puppet_operational_dashboards (>= 1.7.0)
- puppetlabs/influxdb  (>=1.3.1)
- puppetlabs/stdlib (>= 4.5.0 < 9.0.0)
- puppetlabs/concat (>= 1.1.2 < 7.0.0)
- puppetlabs/transition (>= 0.1.0 < 1.0.0)
- puppet/augeasproviders_core (>= 2.1.5 < 4.0.0)
- puppet/augeasproviders_shellvar (>= 1.2.0 < 5.0.0)
- puppetlabs/apt (>= 2.0.0 < 8.0.0)
- puppet-grafana (>= 3.0.0 < 11.0.0)
- puppet-telegraf (>= 2.0.0 < 6.0.0)
- puppetlabs-apt (>= 4.3.0 < 9.0.0)
- puppetlabs-inifile (>= 2.0.0 < 5.0.0)

#### Minimum Hardware requirements


| AWS EC2|Cores| RAM |Disk|
| --- | ----------- | --| --|
| m1.medium | 2 CPU | 4GB Memory | 40GB Disk 

#### OS Restrictions

Puppet_Operations_Appliance will support RHEL / Debian / Ubuntu however due to the additional of PE Client tools in the installation, you are restricted to installing it on a platform with the same OS as the Primary PE Server.

### Beginning with Puppet_Operations_Appliance

Puppet_Operations_Appliance has two main classes for use in the installation:

 - Puppet_Operations_Appliance::exporter - to be applied to all Puppet infrastructure agents - Console node group "PE Infrastructure Agent"
 - Puppet_Operations_Appliance::importer - to be applied to a single node which will be come the Puppet Operations Appliance.

Following the application of these classes to the infrastructure, Puppet Will need to be run on the corresponding agents in the following order:

Infrastructure Agent(s)->Puppet_Operations_Appliance Agent->Infrastructure Agent(s)->Puppet_Operations_Appliance Agent

## Usage
The following outlines the main features of Puppet_Operations_Appliance and how to consume them
### Live Telemetry Display

The Puppet_Operations_Appliance node will host an instance of the [Puppet Operational Dashboard](https://forge.puppet.com/modules/puppetlabs/puppet_operational_dashboards)
 
The Dashboard can be accessed on

**URL:** http://<Puppet_Operations_Appliance-ip\>:3000\
**User:** admin\
**Password:** admin

For advanced configuration and documentation please see [Puppet Operational Dashboard](https://forge.puppet.com/modules/puppetlabs/puppet_operational_dashboards)

### Infrastructure node file and log access	

The Puppet_Operations_Appliance node will, by default, mount `/var/log/`, `/opt/puppetlabs` and `/etc/puppetlabs` from each of the Puppet Enterprise Infrastructure nodes on the Puppet_Operations_Appliance in the following location, as read-only file systems.

`/var/pesupport/<FQDN of Infrastructure node\>/var/log`\
`/var/pesupport/<FQDN of Infrastructure node\>/opt/puppetlabs`\
`/var/pesupport/<FQDN of Infrastructure node\>/etc/puppetlabs`

#### Optional Configuration

The Puppet_Operations_Appliance Class assumes the Puppet_Operations_Appliance server will mount the shared partitions using the IP address Source designated by the "ipaddress" fact. In any deployment should this assertion not be true, it is necessary to set the following parameter to the source IP address of the Puppet_Operations_Appliance Host:

In Hiera 

```
puppet_operations_appliance::exporter::importer_ips:
  - 1.2.3.4
  ```

Console Class Declaration

```
["1.2.3.4"]
```

The Puppet_Operations_Appliance::Exporter class allows for the NFS mounts to be optionally available, to disable existing mounts, or prevent the mounts from installing in the first place set the following parameter:


In Hiera

```
puppet_operations_appliance::exporter::nfsmount: false
```

### PE Client tools

The Puppet_Operations_Appliance node will deploy Puppet Client tools for use by Puppet Enterprise on the Puppet_Operations_Appliance platform, For More information please see the Puppet Enterprise Documentation:

[PE Client tools](https://puppet.com/docs/pe/latest/installing_pe_client_tools.html)

A supplementary task is available to generate an RBAC user and role, so that the credentials may be used provided to Puppet Enterprise Support personnel.  
<br>
#### Creating Support User  
<br>
Run the following task against the Primary Puppet Enterprise Server\
For information on executing PE tasks see the [Puppet Enterprise Documentation](https://puppet.com/docs/pe/latest/tasks_in_pe.html)\
Puppet_Operations_Appliance::supportuser\
When successful the task will return a password, this should be delivered to Puppet Enterprise Support personnel.
<br>
<br>
The Task creates the following user and role:
<br>
<br>

**User:** pesupport 

**Role:** PE Support Role 

The role is intentionally left without permissions, and should be given only the permissions the installing organisation are authorised to grant to Puppet Enterprise Support personnel. For more information on RBAC permissions please see the [Puppet Enterprise Documentation](https://puppet.com/docs/pe/latest/rbac_permissions_intro.html)

### Puppet Enterprise Database Access	

The Puppet_Operations_Appliance Platform has a Postgresql client installed, and is granted certificate based access to all Puppet Enterprise Databases on any pe_postgresl node within the current deployment. The access is limited to the [SELECT](https://www.postgresql.org/docs/11/sql-grant.html) privilege and is therefore READONLY in nature.

To use this function execute the following command from the CLI of the Puppet_Operations_Appliance host

```
psql "host=$(puppet config print server) port=5432 user=puppet_operations_appliance sslmode=verify-full sslcert=$(puppet config print hostcert) sslkey=$(puppet config print hostprivkey) sslrootcert=$(puppet config print localcacert) dbname=<pe_db_name>"
```

Where valid options for <pe_db_name> are:

- pe-rbac 
- pe-puppetdb 
- pe-orchestrator 
- pe-inventory 
- pe-classifier 
- pe-activity

## Uninstallation 

To Uninstall Puppet_Operations_Appliance from your Puppet Enterprise Infrastructure.


 - Remove the following Classification:
Puppet_Operations_Appliance::exporter\
Puppet_Operations_Appliance::importer

 - Add the following classification to the "PE Infrastructure Agent" node group
 Puppet_Operations_Appliance::remove_exporter

  - Remove the following classification to the "PE Infrastructure Agent" node group
 Puppet_Operations_Appliance::remove_exporter

  - Run Puppet on all nodes in "PE Infrastructure Agent" node group

   - Decommission the Puppet_Operations_Appliance platform 


## Limitations
 - The Puppet_Operations_Appliance importer class should only be applied one agent node

## Known Issues

 - Puppet_Operations_Appliance NFS volumes are mounted RW, but exported RO  [26](https://github.com/puppetlabs/Puppet_Operations_Appliance/issues/26)
 
 There is no impact to the end user

## Contributions

For feature development + bug reporting:

 - A Git Issue should exist or be created per feature or Bug
 - Repository should be forked and any changes made by way of PR to the Main Branch
 - PRS should always reference a git issue
  
---

# Supporting Content

### Articles

The [Support Knowledge base](https://support.puppet.com/hc/en-us) is a searchable repository for technical information and how-to guides for all Puppet products.


### Videos

The [Support Video Playlist](https://youtube.com/playlist?list=PLV86BgbREluWKzzvVulR74HZzMl6SCh3S) is a resource of content generated by the support team

This Module has the following specific video content  available:


1. [Puppet Metrics Overview ](https://youtu.be/LiCDoOUS4hg)
2. [Collecting and Displaying Puppet Metrics](https://youtu.be/13sBMQGDqsA)
3. [Interpreting Puppet Metrics](https://youtu.be/09iDO3DlKMQ)

   
   ---

CHANGELOG.md
# Change log

All notable changes to this project will be documented in this file. The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/) and this project adheres to [Semantic Versioning](http://semver.org).

## [v1.0.0](https://github.com/puppetlabs/puppetlabs-puppet_operations_appliance/tree/v1.0.0) (2022-11-09)

[Full Changelog](https://github.com/puppetlabs/puppetlabs-puppet_operations_appliance/compare/v0.3.0...v1.0.0)

### Changed

- Rebrand to Puppet\_Operations\_appliance [\#85](https://github.com/puppetlabs/puppetlabs-puppet_operations_appliance/pull/85) ([MartyEwings](https://github.com/MartyEwings))
- \(SUP-3734\) Switch to Operational Dashboards [\#84](https://github.com/puppetlabs/puppetlabs-puppet_operations_appliance/pull/84) ([MartyEwings](https://github.com/MartyEwings))

## [v0.3.0](https://github.com/puppetlabs/puppetlabs-puppet_operations_appliance/tree/v0.3.0) (2021-12-03)

[Full Changelog](https://github.com/puppetlabs/puppetlabs-puppet_operations_appliance/compare/v0.2.0...v0.3.0)

### Added

- \(SUP-2817\) Consider the distributed architecture of PE\_XL and make log scope configurable [\#81](https://github.com/puppetlabs/puppetlabs-puppet_operations_appliance/pull/81) ([MartyEwings](https://github.com/MartyEwings))

## [v0.2.0](https://github.com/puppetlabs/puppetlabs-puppet_operations_appliance/tree/v0.2.0) (2021-06-28)

[Full Changelog](https://github.com/puppetlabs/puppetlabs-puppet_operations_appliance/compare/v0.1.3...v0.2.0)

### Added

- Make NFS mounts optional [\#66](https://github.com/puppetlabs/puppetlabs-puppet_operations_appliance/pull/66) ([MartyEwings](https://github.com/MartyEwings))

### Fixed

- \(sup-2385\) Resolve unresponsive RSAN node as a result of missing NFS Server [\#64](https://github.com/puppetlabs/puppetlabs-puppet_operations_appliance/pull/64) ([MartyEwings](https://github.com/MartyEwings))

## [v0.1.3](https://github.com/puppetlabs/puppetlabs-puppet_operations_appliance/tree/v0.1.3) (2021-06-10)

[Full Changelog](https://github.com/puppetlabs/puppetlabs-puppet_operations_appliance/compare/0.1.2...v0.1.3)

### Fixed

- Update Readme for Beta Launch [\#59](https://github.com/puppetlabs/puppetlabs-puppet_operations_appliance/pull/59) ([MartyEwings](https://github.com/MartyEwings))

## [0.1.2](https://github.com/puppetlabs/puppetlabs-puppet_operations_appliance/tree/0.1.2) (2021-05-07)

[Full Changelog](https://github.com/puppetlabs/puppetlabs-puppet_operations_appliance/compare/0.1.1...0.1.2)

### Fixed

- \(SUP-2412\) changed privatekey permissions for psql client [\#51](https://github.com/puppetlabs/puppetlabs-puppet_operations_appliance/pull/51) ([MartyEwings](https://github.com/MartyEwings))
- Fix for Issue 49 - New replica installation [\#50](https://github.com/puppetlabs/puppetlabs-puppet_operations_appliance/pull/50) ([MartyEwings](https://github.com/MartyEwings))
- Fix \#46 \(SUP-2384\) [\#47](https://github.com/puppetlabs/puppetlabs-puppet_operations_appliance/pull/47) ([MartyEwings](https://github.com/MartyEwings))
- \(SUP-2386\) updating documentation and dependancies \[SUP-2386\] [\#45](https://github.com/puppetlabs/puppetlabs-puppet_operations_appliance/pull/45) ([MartyEwings](https://github.com/MartyEwings))

## [0.1.1](https://github.com/puppetlabs/puppetlabs-puppet_operations_appliance/tree/0.1.1) (2021-01-18)

[Full Changelog](https://github.com/puppetlabs/puppetlabs-puppet_operations_appliance/compare/0.1.0...0.1.1)

### Fixed

- Metadata and spelling corrections [\#42](https://github.com/puppetlabs/puppetlabs-puppet_operations_appliance/pull/42) ([MartyEwings](https://github.com/MartyEwings))

## [0.1.0](https://github.com/puppetlabs/puppetlabs-puppet_operations_appliance/tree/0.1.0) (2021-01-15)

[Full Changelog](https://github.com/puppetlabs/puppetlabs-puppet_operations_appliance/compare/9776f5017215976a4bb2da7565482e04ef4b065b...0.1.0)

### Added

- Documentation and release prep [\#41](https://github.com/puppetlabs/puppetlabs-puppet_operations_appliance/pull/41) ([MartyEwings](https://github.com/MartyEwings))
- remove mocha deprecation warnings [\#39](https://github.com/puppetlabs/puppetlabs-puppet_operations_appliance/pull/39) ([MartyEwings](https://github.com/MartyEwings))
- {WIP} remove exporter class state from infra nodes [\#36](https://github.com/puppetlabs/puppetlabs-puppet_operations_appliance/pull/36) ([MartyEwings](https://github.com/MartyEwings))
- Feature - PE Client tools Deployment [\#12](https://github.com/puppetlabs/puppetlabs-puppet_operations_appliance/pull/12) ([MartyEwings](https://github.com/MartyEwings))
- Morgan lillie [\#11](https://github.com/puppetlabs/puppetlabs-puppet_operations_appliance/pull/11) ([morgan-lillie](https://github.com/morgan-lillie))
- W.I.P. PR2: New functions created to query for each item - calling of functions in importer.pp [\#10](https://github.com/puppetlabs/puppetlabs-puppet_operations_appliance/pull/10) ([liamjohnsexton](https://github.com/liamjohnsexton))
- add codeownerfiles [\#9](https://github.com/puppetlabs/puppetlabs-puppet_operations_appliance/pull/9) ([MartyEwings](https://github.com/MartyEwings))

### Fixed

- https://github.com/MartyEwings/RSAN/issues/6 [\#7](https://github.com/puppetlabs/puppetlabs-puppet_operations_appliance/pull/7) ([MartyEwings](https://github.com/MartyEwings))



\* *This Changelog was automatically generated by [github_changelog_generator](https://github.com/github-changelog-generator/github-changelog-generator)*


spec/spec_helper.rb
# frozen_string_literal: true

require 'puppetlabs_spec_helper/module_spec_helper'
require 'rspec-puppet-facts'

require 'spec_helper_local' if File.file?(File.join(File.dirname(__FILE__), 'spec_helper_local.rb'))

include RspecPuppetFacts

default_facts = {
  puppetversion: Puppet.version,
  facterversion: Facter.version,
}

default_fact_files = [
  File.expand_path(File.join(File.dirname(__FILE__), 'default_facts.yml')),
  File.expand_path(File.join(File.dirname(__FILE__), 'default_module_facts.yml')),
]

default_fact_files.each do |f|
  next unless File.exist?(f) && File.readable?(f) && File.size?(f)

  begin
    default_facts.merge!(YAML.safe_load(File.read(f), [], [], true))
  rescue => e
    RSpec.configuration.reporter.message "WARNING: Unable to load #{f}: #{e}"
  end
end

# read default_facts and merge them over what is provided by facterdb
default_facts.each do |fact, value|
  add_custom_fact fact, value
end

RSpec.configure do |c|
  c.default_facts = default_facts
  c.hiera_config = File.expand_path(File.join(__dir__, 'fixtures/hiera.yaml'))
  c.before :each do
    # set to strictest setting for testing
    # by default Puppet runs at warning level
    Puppet.settings[:strict] = :warning
    Puppet.settings[:strict_variables] = true
  end
  c.filter_run_excluding(bolt: true) unless ENV['GEM_BOLT']
  c.after(:suite) do
  end
end

# Ensures that a module is defined
# @param module_name Name of the module
def ensure_module_defined(module_name)
  module_name.split('::').reduce(Object) do |last_module, next_module|
    last_module.const_set(next_module, Module.new) unless last_module.const_defined?(next_module, false)
    last_module.const_get(next_module, false)
  end
end

# 'spec_overrides' from sync.yml will appear below this line


spec/classes/node_spec.rb
require 'spec_helper'

describe 'rsan::node' do
  on_supported_os.each do |os, os_facts|
    context "on #{os}" do
      let(:facts) { os_facts }

      it { is_expected.to compile }
    end
  end
end


spec/classes/rsan_spec.rb
# frozen_string_literal: true

require 'spec_helper'

describe 'rsan' do
  on_supported_os.each do |os, os_facts|
    context "on #{os}" do
      let(:facts) { os_facts }

      it { is_expected.to compile }
    end
  end
end


tasks/init.sh
#!/bin/sh

# Puppet Task Name: rsan
#
# Learn more at: https://puppet.com/docs/bolt/0.x/writing_tasks.html#ariaid-title11
#

#
# shellcheck disable=SC2046

declare PT_rsannode
rsannode=$PT_rsannode

mom=$(hostname -f)
rsanip=$(curl -X GET http://localhost:8080/pdb/query/v4/facts/ipaddress   --data-urlencode "query=[\"=\", \"certname\", \"$rsannode\"]" | awk -v RS='([0-9]+\\.){3}[0-9]+' 'RT{print RT}')
infranodes=($(puppet query "nodes[certname] { resources { type = 'Class' and title = 'Puppet_enterprise::Profile::Master' or type = 'Class' and title = 'puppet_enterprise::profile::primary_master_replica'} and certname !='$(hostname -f)'  }" --urls http://localhost:8080 | sed -e 's/[{}]/''/g' |      awk -v k="text" '{n=split($0,a,":"); for (i=1; i<=n; i++) print a[i+1]}' | sed '/^$/d'))

#inodes=$(printf "$mom ,[%s] " "${infranodes[@]/%/, 8140}")

inodes=$(printf ",[%s]" "${infranodes[@]/%/, 8140}")
classinodes=$(echo "[""\""$mom"\""$inodes"]")



if [ -e "/etc/sysconfig/pe-puppetserver" ] # Test to confirm this is a Puppetserver
then
  echo "Puppet master node detected"   #Log Line to StdOut for the Console







############## Create pesupport User and role for use in client tools##########
###############################################################################

password=$(/opt/puppetlabs/puppet/bin/openssl rand -base64 32) #Set Random Password for Support Client tools account

curl -X POST -H 'Content-Type: application/json' --cert $(puppet config print hostcert) --key $(puppet config print hostprivkey) --cacert $(puppet config print localcacert) https://$(hostname -f):4433/rbac-api/v1/users -d "{\"login\":\"pesupport\",\"email\":\"support@puppet.com\",\"role_ids\": [],\"display_name\":\"Puppet Enterprise Support\", \"password\": \"$password\"}"

echo "password for pesupport account is $password"

peusersid=$(curl -X get -H 'Content-Type: application/json' --cert $(puppet config print hostcert) --key $(puppet config print hostprivkey) --cacert $(puppet config print localcacert) https://$(hostname -f):4433/rbac-api/v1/users|sed -e 's/[{}]/''/g' |      awk -v k="text" '{n=split($0,a,","); for (i=1; i<=n; i++) {if (a[i]=="\"login\":\"pesupport\""){ print a[i+2]}}}' | awk '{split($0,a,"\""); print a[4]}') #get the SID of the new user to use in adding a role


curl -X POST -H 'Content-Type: application/json' --cert $(puppet config print hostcert) --key $(puppet config print hostprivkey) --cacert $(puppet config print localcacert) https://$(hostname -f):4433/rbac-api/v1/roles -d "{\"description\":\"Puppet Enterprise Support user role\",\"display_name\":\"PE Suport Role\",\"user_ids\":[\"$peusersid\"],\"group_ids\":[],\"permissions\":[{\"object_type\":\"node_groups\",\"action\":\"modify_children\",\"instance\":\"*\"},{\"object_type\":\"node_groups\",\"action\":\"set_environment\",\"instance\":\"*\"},{\"object_type\":\"node_groups\",\"action\":\"view\",\"instance\":\"*\"},{\"object_type\":\"puppet_agent\",\"action\":\"run\",\"instance\":\"*\"},{\"object_type\":\"environment\",\"action\":\"deploy_code\",\"instance\":\"*\"},{\"object_type\":\"nodes\",\"action\":\"view_data\",\"instance\":\"*\"},{\"object_type\":\"node_groups\",\"action\":\"edit_config_data\",\"instance\":\"*\"},{\"object_type\":\"orchestrator\",\"action\":\"view\",\"instance\":\"*\"}]}"  #create role add user




############## create and classify node group for rsan node 
###############################################################################

curl -X PUT -H 'Content-Type: application/json' --cert $(puppet config print hostcert) --key $(puppet config print hostprivkey) --cacert $(puppet config print localcacert) https://$(hostname -f):4433/classifier-api/v1/groups/72fef894-4a71-473a-9e54-c94a2805ddd9 -d "{\"name\":\"RemoteSupportAccess\",\"parent\":\"00000000-0000-4000-8000-000000000000\",\"classes\": {\"rsan\":{\"pdb\":\"$mom\",\"infranode\":$classinodes}}}"

echo "RemoteSupportAccess group created"

curl -X POST -H 'Content-Type: application/json' --cert $(puppet config print hostcert) --key $(puppet config print hostprivkey) --cacert $(puppet config print localcacert) https://$(hostname -f):4433/classifier-api/v1/groups/72fef894-4a71-473a-9e54-c94a2805ddd9/pin -d {\"nodes\":[\"$rsannode\"]}

echo "$rsannode has been pined to group"



#######master node classify and pin

curl -X PUT -H 'Content-Type: application/json' --cert $(puppet config print hostcert) --key $(puppet config print hostprivkey) --cacert $(puppet config print localcacert) https://$(hostname -f):4433/classifier-api/v1/groups/7434a707-7104-4142-85b8-5404768a841a -d "{\"name\":\"RSANMaster\",\"parent\":\"00000000-0000-4000-8000-000000000000\",\"classes\": {\"rsan::master\":{\"rsanip\":\"$rsanip\"}}}"

echo "RSANMaster  group created"

curl -X POST -H 'Content-Type: application/json' --cert $(puppet config print hostcert) --key $(puppet config print hostprivkey) --cacert $(puppet config print localcacert) https://$(hostname -f):4433/classifier-api/v1/groups/7434a707-7104-4142-85b8-5404768a841a/pin -d {\"nodes\":[\"$mom\"]}

echo "$mom has been pined to group"

######infra nodes classify and pin

curl -X PUT -H 'Content-Type: application/json' --cert $(puppet config print hostcert) --key $(puppet config print hostprivkey) --cacert $(puppet config print localcacert) https://$(hostname -f):4433/classifier-api/v1/groups/8882255f-26d9-47cf-a6e4-5b9075d30703 -d "{\"name\":\"RSANInfra\",\"parent\":\"00000000-0000-4000-8000-000000000000\",\"rule\": [\"=\", [\"trusted\",\"extensions\", \"pp_auth_role\"], \"pe_compiler\"],\"classes\": {\"rsan::infrastructure\":{\"rsanip\":\"$rsanip\"}}}"


echo "RSANInfra  group created"
echo "PE_Compilers will automatically be classified, please pin any Replicas or Legacy compilers to this group manually"


else
  echo  "Not a Puppet Enterprise Master node, exiting"

fi


tasks/supportuser.sh
#!/bin/sh
# shellcheck disable=2046

# Puppet Task Name: supportuser
#
password=$(/opt/puppetlabs/puppet/bin/openssl rand -base64 32) #Set Random Password for Support Client tools account

curl -X POST -H 'Content-Type: application/json' --cert $(puppet config print hostcert) --key $(puppet config print hostprivkey) --cacert $(puppet config print localcacert) https://$(hostname -f):4433/rbac-api/v1/users -d "{\"login\":\"pesupport\",\"email\":\"support@puppet.com\",\"role_ids\": [],\"display_name\":\"Puppet Enterprise Support\", \"password\": \"$password\"}"

echo "{\"status\":\"created\",\"password\":\"$password\"}"

peusersid=$(curl -X get -H 'Content-Type: application/json' --cert $(puppet config print hostcert) --key $(puppet config print hostprivkey) --cacert $(puppet config print localcacert) https://$(hostname -f):4433/rbac-api/v1/users|sed -e 's/[{}]/''/g' |      awk -v k="text" '{n=split($0,a,","); for (i=1; i<=n; i++) {if (a[i]=="\"login\":\"pesupport\""){ print a[i+2]}}}' | awk '{split($0,a,"\""); print a[4]}') #get the SID of the new user to use in adding a role


curl -X POST -H 'Content-Type: application/json' --cert $(puppet config print hostcert) --key $(puppet config print hostprivkey) --cacert $(puppet config print localcacert) https://$(hostname -f):4433/rbac-api/v1/roles -d "{\"description\":\"Puppet Enterprise Support user role\",\"display_name\":\"PE Suport Role\",\"user_ids\":[\"$peusersid\"],\"group_ids\":[],\"permissions\":[{\"object_type\":\"node_groups\",\"action\":\"view\",\"instance\":\"*\"}]}"  #create role add user


tasks/supportsession.sh
#!/bin/sh

# Puppet Task Name: supportsession
#
# This is where you put the shell code for your task.
#
# You can write Puppet tasks in any language you want and it's easy to
# adapt an existing Python, PowerShell, Ruby, etc. script. Learn more at:
# https://puppet.com/docs/bolt/0.x/writing_tasks.html
#
# Puppet tasks make it easy for you to enable others to use your script. Tasks
# describe what it does, explains parameters and which are required or optional,
# as well as validates parameter type. For examples, if parameter "instances"
# must be an integer and the optional "datacenter" parameter must be one of
# portland, sydney, belfast or singapore then the .json file
# would include:
#   "parameters": {
#     "instances": {
#       "description": "Number of instances to create",
#       "type": "Integer"
#     },
#     "datacenter": {
#       "description": "Datacenter where instances will be created",
#       "type": "Enum[portland, sydney, belfast, singapore]"
#     }
#   }
# Learn more at: https://puppet.com/docs/bolt/0.x/writing_tasks.html#ariaid-title11
#


lib/puppet/util/nodegroups_api.rb
require 'puppet/network/http_pool'
require 'json'

class Puppet::Util::Nodegroups_api
  def initialize
    pe_agent_config = JSON.parse(Facter::Util::Resolution.exec('sudo -u root /opt/puppetlabs/bin/puppet config print --render-as json'))

    settings_file = "#{pe_agent_config['confdir']}/classifier.yaml"

    begin
      nc_settings = YAML.load_file(settings_file)
      nc_settings = nc_settings.first if nc_settings.class == Array
    rescue
      fail "Could not find file #{settings_file}"
    else
      cl_server       = nc_settings['server'] || pe_agent_config['server']
      cl_port         = nc_settings['port']   || 4433
      @classifier_url = "https://#{cl_server}:#{cl_port}/classifier-api"
      @token          = nc_settings['token']
      unless cl_server == pe_agent_config['certname']
        remote_client = "#{Facter.value('fqdn')} (#{Facter.value('ipaddress')})"
        Puppet.debug("Managing node_group remotely from #{remote_client}")
      end
      Puppet.debug("classifier_url: #{@classifier_url}")

      unless @token and ! @token.empty?
        @ca_certificate_path = nc_settings['localcacert'] || pe_agent_config['localcacert']
        @certificate_path    = nc_settings['hostcert']    || pe_agent_config['hostcert']
        @private_key_path    = nc_settings['hostprivkey'] || pe_agent_config['hostprivkey']
      end
    end
  end

  def get_groups
    res = do_https('v1/groups', 'GET')
    if res.code.to_i != 200
      error_msg(res)
      fail('Unable to get node_group list')
    else
      JSON.parse(res.body)
    end
  end

  private

  def do_https(endpoint, method = 'post', data = {})
    url  = "#{@classifier_url}/#{endpoint}"
    uri  = URI(url)
    http = Net::HTTP.new(uri.host, uri.port)

    unless @token and ! @token.empty?
      Puppet.debug('Using SSL authentication')
      http.use_ssl     = true
      http.cert        = OpenSSL::X509::Certificate.new(File.read @certificate_path)
      http.key         = OpenSSL::PKey::RSA.new(File.read @private_key_path)
      http.ca_file     = @ca_certificate_path
      http.verify_mode = OpenSSL::SSL::VERIFY_CLIENT_ONCE
    else
      Puppet.debug('Using token authentication')
      http.use_ssl     = true
      http.verify_mode = OpenSSL::SSL::VERIFY_NONE
    end

    req              = Net::HTTP.const_get(method.capitalize).new(uri.request_uri)
    req.body         = data.to_json
    req.content_type = 'application/json'

    # If using token
    req['X-Authentication'] = @token if @token

    begin
      res = http.request(req)
    rescue Exception => e
      fail(e.message)
      debug(e.backtrace.inspect)
    else
      res
    end
  end
  def error_msg(res)
    json = JSON.parse(res.body)
    kind = json['kind']
    msg  = json['msg']
    Puppet.err %(node_manager failed with error type '#{kind}': #{msg})
    Puppet.debug("Response code: #{res.code}")
    Puppet.debug("Response message: #{res.body}")
  end
end


lib/puppet/functions/nodegroups.rb
require 'json'

Puppet::Functions.create_function(:nodegroups) do
  dispatch :nodegroups do
    required_param 'String', :pe_server
    optional_param 'String', :node_name
  end

  def nodegroups(pe_server, node_name = '')
    puppet_settings = {
      'confdir'     => '/etc/puppetlabs/puppet',
      'certname'    => pe_server,
      'server'      => pe_server,
      'localcacert' => '/etc/puppetlabs/puppet/ssl/certs/ca.pem',
      'hostcert'    => "/etc/puppetlabs/orchestration-services/ssl/#{pe_server}.cert.pem",
      'hostprivkey' => "/etc/puppetlabs/orchestration-services/ssl/#{pe_server}.private_key.pem",
    }

    settings_file = "#{puppet_settings['confdir']}/classifier.yaml"

    begin
      nc_settings = YAML.load_file(settings_file)
      nc_settings = nc_settings.first if nc_settings.class == Array
    rescue
      raise "Could not find file #{settings_file}"
    else
      cl_server       = nc_settings['server'] || puppet_settings['server']
      cl_port         = nc_settings['port']   || 4433
      @classifier_url = "https://#{cl_server}:#{cl_port}/classifier-api"
      @token          = nc_settings['token']
      unless cl_server == puppet_settings['certname']
        remote_client = "#{Facter.value('fqdn')} (#{Facter.value('ipaddress')})"
        Puppet.debug("Managing node_group remotely from #{remote_client}")
      end
      Puppet.debug("classifier_url: #{@classifier_url}")

      unless @token && !@token.empty?
        @ca_certificate_path = nc_settings['localcacert'] || puppet_settings['localcacert']
        @certificate_path    = nc_settings['hostcert']    || puppet_settings['hostcert']
        @private_key_path    = nc_settings['hostprivkey'] || puppet_settings['hostprivkey']
      end
    end

    res = do_https('v1/groups', 'GET')
    if res.code.to_i != 200
      error_msg(res)
      raise('Unable to get node_group list')
    else
      groups = JSON.parse(res.body)
    end

    # When querying a specific group
    if node_name.length.zero?
      hashify_group_array(groups)
    else
      # Assuming there is only one group by the name
      hashify_group_array(
        groups.select { |g| g['name'] == node_name },
      )
    end
  end

  def do_https(endpoint, method = 'post', data = {})
    url  = "#{@classifier_url}/#{endpoint}"
    uri  = URI(url)
    http = Net::HTTP.new(uri.host, uri.port)

    if @token && !@token.empty?
      Puppet.debug('Using token authentication')
      http.use_ssl     = true
      http.verify_mode = OpenSSL::SSL::VERIFY_NONE
    else
      Puppet.debug('Using SSL authentication')
      http.use_ssl     = true
      http.cert        = OpenSSL::X509::Certificate.new(File.read(@certificate_path))
      http.key         = OpenSSL::PKey::RSA.new(File.read(@private_key_path))
      http.ca_file     = @ca_certificate_path
      http.verify_mode = OpenSSL::SSL::VERIFY_CLIENT_ONCE
    end

    req              = Net::HTTP.const_get(method.capitalize).new(uri.request_uri)
    req.body         = data.to_json
    req.content_type = 'application/json'

    # If using token
    req['X-Authentication'] = @token if @token

    begin
      res = http.request(req)
    rescue Exception => e
      raise(e.message)
      debug(e.backtrace.inspect)
    else
      res
    end
  end

  def error_msg(res)
    json = JSON.parse(res.body)
    kind = json['kind']
    msg  = json['msg']
    Puppet.err %(node_manager failed with error type '#{kind}': #{msg})
    Puppet.debug("Response code: #{res.code}")
    Puppet.debug("Response message: #{res.body}")
  end

  def hashify_group_array(group_array)
    hashified = {}

    group_array.each do |group|
      hashified[group['name']] = group
    end

    hashified
  end
end
